bittide/src/Bittide/ClockControl/CallistoSw.hs

```diff
@@ -0,0 +1,163 @@
+-}
+callistoSwClockControl ::
+  forall nLinks eBufBits dom margin framesize.
+  ( KnownDomain dom
+  , KnownNat nLinks
+  , KnownNat eBufBits
+  , 1 <= nLinks
+  , 1 <= eBufBits
+  , nLinks + eBufBits <= 32
+  , 1 <= framesize
+  , 1 <= DomainPeriod dom
+  ) =>
+  -- | CPU clock
+  Clock dom ->
+  -- | CPU reset
+  Reset dom ->
+  -- | CPU enable
+  Enable dom ->
+  -- | Clock control config
+  SwControlConfig dom margin framesize ->
+  -- | Availability mask
+  Signal dom (BitVector nLinks) ->
+  -- | Diff counters
+  Vec nLinks (Signal dom (RelDataCount eBufBits)) ->
+  Signal dom (CallistoResult nLinks)
+callistoSwClockControl clk rst ena (SwControlConfig (reframe :: Signal dom Bool) mgn fsz) mask ebs =
+  hwSeqX callistoSwIla callistoResult
+ where
+  callistoResult =
+    CallistoResult
+      <$> ccData.clockMod
+      <*> ccData.stabilityIndications
+      <*> ccData.allStable
+      <*> ccData.allSettled
+      <*> resultRfs
+
+  resultRfs = fromMaybe Done <$> debugData.reframingState
+
+  callistoSwIla :: Signal dom ()
+  callistoSwIla =
+    setName @"callistoSwIla"
+      $ ila
+        ( ilaConfig
+            $ "trigger_0"
+            :> "capture_0"
+            :> "probe_updatePeriod"
+            :> "probe_updatePeriodMin"
+            :> "probe_updatePeriodMax"
+            :> Nil
+        )
+          { depth = D16384
+          }
+        clk
+        (unsafeToActiveLow rst)
+        capture
+        debugData.updatePeriod
+        debugData.updatePeriodMin
+        debugData.updatePeriodMax
+
+  debugRegisterCfg = DebugRegisterCfg <$> reframe
+
+  capture = isRising clk rst ena False (isJust <$> ccData.clockMod)
+
+  (_, (ccData, debugData)) =
+    toSignals
+      ( circuit $ \jtag -> do
+          [wbClockControl, wbDebug, wbDummy] <-
+            withClockResetEnable clk rst ena $ processingElement peConfig -< jtag
+          idleSink -< wbDummy
+          [ccd0, ccd1] <-
+            cSignalDupe
+              <| withClockResetEnable
+                clk
+                rst
+                ena
+                (clockControlWb mgn fsz mask ebs)
+              -< wbClockControl
+          cm <- cSignalMap clockMod -< ccd0
+          dbg <-
+            withClockResetEnable clk rst enableGen
+              $ debugRegisterWb debugRegisterCfg
+              -< (wbDebug, cm)
+          idC -< (ccd1, dbg)
+      )
+      (pure $ JtagIn low low low, (pure (), pure ()))
+  (iMem, dMem) =
+    $( do
+        root <- runIO $ findParentContaining "cabal.project"
+        let
+          elfDir = root </> firmwareBinariesDir "riscv32imc-unknown-none-elf" Release
+          elfPath = elfDir </> "clock-control"
+          iSize = 64 * 1024 -- 64 KB
+          dSize = 64 * 1024 -- 64 KB
+        memBlobsFromElf BigEndian (Just iSize, Just dSize) elfPath Nothing
+     )
+  peConfig =
+    PeConfig
+      (0b100 :> 0b010 :> 0b110 :> 0b111 :> 0b001 :> Nil)
+      (Reloadable $ Blob iMem)
+      (Reloadable $ Blob dMem)
```

Comment:
Since there is only one clock domain here, can we try to get away with using `exposeClockResetEnable` combined with `hasClock`, `hasReset`, `hasEnable`? ping me if it's not trivial for you to do :)